Update package to the latest streamly master

adithyaov opened the pull request.

adithyaov committed:
Update package to the latest streamly master

hasufell commented on streamly-bytestring.cabal:
```diff
@@ -31,8 +31,11 @@ library
+    , bytestring    == 0.11.0.*
+                 || == 0.11.1.*
+                 || == 0.11.2.*
+                 || == 0.11.3.*
```
```suggestion
                 || ( >= 0.11.3.1 && < 0.11.4 )
```

adithyaov committed:
Un-support bytestring-0.11.3.0

bytestring-0.11.3.0 causes panics and other issues on windows.
the current constraint doesn't make sense from a PVP point of view
either.

* https://gitlab.haskell.org/ghc/ghc/-/issues/21179
* https://github.com/haskell/bytestring/pull/500

Co-authored-by: Julian Ospald <hasufell@posteo.de>

adithyaov committed:
Update the stack configuration

adithyaov committed:
Bump the package version and update the changelog

adithyaov committed:
Add GHC 9.4.2 and 9.2.4 to the CI

adithyaov committed:
Remove Azure CI pipelines

jack-davies commented:
Hey folks, thanks for your work on this.

I'm currently trying to migrate a number of packages to GHC 9 with the stackage `LTS-20` release, but unfortunately the strict `bytestring <=0.11.3.0` bound in **streamly-bytestring** is blocking the upgrade path as the snapshot contains `0.11.3.1`.

This is additionally causing **streamly-bytestring** to be [excluded](https://github.com/commercialhaskell/stackage/blob/59e4012189edab99d5693d9a2cba0c00c5d6e6b8/build-constraints.yaml#L7005) from the latest stackage snapshot.

Is this PR ready to go? Alternately, would you be happy to accept a small patch to update just the bounds of **bytestring**?

harendra-kumar commented:
@jack-davies this PR is to adapt this package to changes in streamly-0.9.0 which is yet to be released. It may not help you.

Would it be sufficient to relax the upper bound of streamly-bytestring on Hackage? @adithyaov can you please look into it and unblock @jack-davies ? We will be happy to accept a PR or patch.

harendra-kumar commented:
@adithyaov can you please update this PR to sync it with latest changes in streamly master branch. We can then merge it to prepare for the next release.

jack-davies commented:
> Would it be sufficient to relax the upper bound of streamly-bytestring on Hackage?

Yes, this would be sufficient. Thanks for the quick response!

I've got a PR at https://github.com/psibi/streamly-bytestring/pull/28.

adithyaov committed:
Update the streamly commit in the project file

adithyaov commented on src/Streamly/External/ByteString.hs:
```diff
@@ -14,38 +14,67 @@ where
-  where
-    startPtr = Ptr addr# `plusPtr` off
-    endPtr = startPtr `plusPtr` len
+toArray (BS (ForeignPtr addr# _) _)
+    | Ptr addr# == nullPtr = Array nilArrayContents 0 0
+toArray (BS (ForeignPtr addr# (PlainPtr marr#)) len) =
+    let off = I# (addr# `minusAddr#` mutableByteArrayContents# marr#)
+     in Array (MutableByteArray marr#) off (off + len)
+toArray (BS fptr len) =
+    unsafeInlineIO
+        $ withForeignPtr fptr $ Unfold.fold (Array.writeN len) generator
+
+    where
+
+    generator =
+        Unfold.mkUnfoldrM
+            (\ptr -> flip StreamD.Yield (ptr `plusPtr` 1) <$> peek ptr)
```
We can use `fromPtr` here maybe? That will result in a pinned mutable byte array.

adithyaov committed:
Fix build with the latest streamly master

adithyaov committed:
Fix test-suite

adithyaov committed:
Fix bench-suite

adithyaov committed:
Remove explicit dependencies from cabal.project.user

adithyaov committed:
Remove streamly from the dependency list

harendra-kumar commented on Changelog.md:
```diff
@@ -1,5 +1,10 @@
 # Changelog for streamly-bytestring
 
+## 0.2.0 (Unreleased)
```
Add release date.

harendra-kumar commented on src/Streamly/External/ByteString.hs:
```diff
@@ -14,38 +14,67 @@ where
+import qualified Streamly.Data.Array as Array
+import qualified Streamly.Internal.Data.Unboxed as Unboxed
+import qualified Streamly.Internal.Data.Unfold as Unfold
+import qualified Streamly.Internal.Data.Stream.StreamD as StreamD
```
Can we remove the Internal in all these imports? If there are any really internal APIs required import those explicitly.

harendra-kumar commented on src/Streamly/External/ByteString.hs:
```diff
@@ -14,38 +14,67 @@ where
+     in Array (MutableByteArray marr#) off (off + len)
+toArray (BS fptr len) =
+    unsafeInlineIO
+        $ withForeignPtr fptr $ Unfold.fold (Array.writeN len) generator
```
We have a `fromPtrN` stream generator in streamly which can be used here.

adithyaov replied:
I'm unable to reason with this now. I'm leaving this as it is.

harendra-kumar commented on src/Streamly/External/ByteString.hs:
```diff
@@ -14,38 +14,67 @@ where
 
 -- | Convert a 'ByteString' to an array of 'Word8'. This function unwraps the
 -- 'ByteString' and wraps it with 'Array' constructors and hence the operation
 -- is performed in constant time.
```
It can be done in constant time only for GHC allocated memory. For foreign allocator allocated memory there is a copy involved.

harendra-kumar commented on src/Streamly/External/ByteString/Lazy.hs:
```diff
@@ -9,20 +9,25 @@ module Streamly.External.ByteString.Lazy
+import qualified Streamly.Data.Array as Array
+import qualified Streamly.Data.Unfold as Unfold
+import qualified Streamly.Data.Stream as Stream
+import qualified Streamly.Internal.Data.Stream.StreamD.Eliminate as Stream
```
This can be removed. In fact there are warnings below.

harendra-kumar commented on stack.yaml:
```diff
@@ -1,11 +1,22 @@
-resolver: lts-18.18
+resolver: lts-19.25
```
We can change this lts-20.14

harendra-kumar commented on stack.yaml:
```diff
@@ -1,11 +1,22 @@
+- github: composewell/streamly
+  commit: "master"
+  subdirs:
+    - core
```
These can be removed.

harendra-kumar commented on streamly-bytestring.cabal:
```diff
@@ -31,8 +31,11 @@ library
+    , bytestring    == 0.11.0.*
+                 || == 0.11.1.*
+                 || == 0.11.2.*
+                 || ( >= 0.11.3.1 && < 0.11.4 )
```
0.11.4.0 is not supported? That's the latest version.
Is there is problem with 0.11.3.0? That version is excluded.

harendra-kumar commented on test/Main.hs:
```diff
@@ -5,41 +5,42 @@ module Main where
 word8List :: Int -> IO [Word8]
 word8List l =
-    S.toList $
+    S.fold Fold.toList $
```
These could be S.toList now.

adithyaov replied:
I'm leaving this as is

adithyaov committed:
Update the date in the changelog

adithyaov committed:
Support bytestring 0.11.4.*

adithyaov committed:
Update the stack file

adithyaov committed:
Make internal imports explicit and update the docs

adithyaov committed:
Add the Werror flag in the cabal components

adithyaov committed:
Remove unused imports

adithyaov committed:
Rename read to reader, deprecate read, and update README

adithyaov committed:
Unsupport GHC < 8.6.5

adithyaov committed:
Enable building benchmarks in the CI

adithyaov committed:
Remove Werror from the cabal file and add it only in the CIs

adithyaov committed:
Update the changelog to indicate streamly-core

harendra-kumar commented on streamly-bytestring.cabal:
```diff
@@ -28,7 +28,7 @@ library
   hs-source-dirs:
       src
-  ghc-options: -Wall -O2
+  ghc-options: -Werror -Wall -O2
```
This is not good at all. Should not add -Werror for all the users of the library. We should use a separate cabal.project file to do it only in CI.

adithyaov replied:
This was fixed in another PR.

harendra-kumar reviewed (approved):
Assuming the other PR fixes the -Werror flag

adithyaov committed:
Add deprecation warnings
